### PR TITLE
feat(playwright-reporter-llm)!: three-layer network schema with priority retention

### PR DESCRIPTION
**Pull request by therockstorm**
## Summary

- Replaces the flat `NetworkRequest[]` (schemaVersion 2) with a three-layer `NetworkReport` (schemaVersion 3): instances (what happened), groups (shape-deduped via fingerprint), and content-addressed bodies.
- Adds a concrete tiered priority score (5xx > connect/DNS/TLS failures > 4xx > successful xhr/fetch > aborted > unknown > other-known > static asset), duplicate-stride sampling (first 3 then 1-in-10), and strict-`<` eviction so high-signal requests survive under pressure.
- Adds `NetworkSummary` counters that satisfy an exact accounting invariant (`observedInstances === retained + droppedByFilter + droppedByGroupCap + droppedByInstanceCap + suppressedAsDuplicate + evictedAfterAdmission`); the reporter catches invariant violations and substitutes an empty report + `GlobalError` instead of crashing.
- Timeline network entries now carry `networkId` so agents can resolve to full per-request detail.

## Why

In a representative failing trace the old retention policy leaked ~32 static JS assets past the filter and let ~138 polling duplicates crowd out ~30 genuinely distinct requests within the 200-entry cap. Consumers had no way to tell a quiet attempt from a saturated one, and timeline entries value-copied fields with no link back to the underlying request.

Plan: `~/.claude/plans/linked-finding-tide.md`.

User confirmed no backward-compat constraints, so this ships as a clean v3 break.

## Test plan

- [x] `nx run playwright-reporter-llm:lint` — clean
- [x] `nx run playwright-reporter-llm:test` — 186 passing (includes new fingerprint-collision, oldest-firstOffset eviction, forward-only redirect linking, saturation, invariant-violation tests)
- [x] `nx run playwright-reporter-llm:build` — typechecks under `exactOptionalPropertyTypes`
- [x] Root `node --run ci:check` — format / cspell / syncpack / embed / knip / oxlint / markdownlint / architecture all clean
- [x] Regression tests mutation-verified against the pre-fix bugs (redirect back-linking, Map-order group eviction, delimiter collision in `hashShape`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/clipboardhealth/core-utils/pull/588" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: packages/playwright-reporter-llm/src/lib/internal/networkBuilder.test.ts [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+
+      // Force group eviction by admitting a tier-10 with a brand-new shape, saturating again.
+      // First, saturate groups by adding a tier-10 that evicts the oldest tier-8.
+      builder.admit(
+        makeObs({ url: "https://api.example.com/evictor", status: 503 }, { offsetMs: 2000 }),
+      );
+
+      const report = builder.finalize();
+      expect(report.summary.retainedGroups).toBeLessThanOrEqual(GROUPS_CAP);
+      assertInvariants(report.summary);
```
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/networkBuilder.test.ts]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/1835fdbae9bfb6950671bbc7548772d73d7e83cf. Agreed — the test filled groups cap with same-tier (400) shapes and the target admits were dropped at the group-cap gate before the sampling counter could build up. Replaced with a group-eviction-cleanup test that actually exercises the cascade path: target admitted past the sampling threshold (13×), then a tier-10 shape forces group-cap eviction; test asserts the target's 4 retained instances are counted in `instancesEvictedAfterAdmission` and the invariants balance. Note: the true "re-admit shape after eviction with counter reset" path is structurally unreachable in this builder — eviction requires lowest priority, re-admission requires beating something, so a shape can't be both at cap-saturation. Documenting the limit rather than shipping a misleading test.

File: packages/playwright-reporter-llm/src/lib/internal/networkBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+        if (candidate?.url === redirectToUrl && candidate.redirectFromId === undefined) {
+          target = candidate;
+          break;
+        }
+      }
+      if (target) {
+        instance.redirectToId = target.id;
+        target.redirectFromId = instance.id;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/networkBuilder.ts:554]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/1835fdbae9bfb6950671bbc7548772d73d7e83cf. Swapped `finalize()` to call `sortInstances()` BEFORE `annotateRedirects()` so the redirect walk iterates in temporal (offsetMs, instanceSequence) order rather than parser/admission order. Added a regression test that admits the redirect target at offsetMs 500 BEFORE the redirect source at offsetMs 100 — mutation-verified it fails under the old ordering.

File: packages/playwright-reporter-llm/src/lib/internal/networkProcessing.ts [hunk 3]
```diff
@@ -157,102 +144,178 @@ function readTraceResourceBody(
...
+  const cappedContent = capNetworkBody(rawContent);
+  const truncated = cappedContent !== rawContent;
+
+  const body: NetworkObservationBody = {
+    content: cappedContent,
+    truncated,
+    fingerprint: hashBody(cappedContent),
+  };
+  if (contentType !== undefined) {
+    body.contentType = contentType;
```
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/networkProcessing.ts:157]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/1835fdbae9bfb6950671bbc7548772d73d7e83cf. `hashBody` now takes `(content, contentType?)` and serializes both via `JSON.stringify([content, contentType ?? null])`. Call site in `networkProcessing.readTraceResourceBody` passes the MIME type too, so two bodies with identical bytes but different content types hash to different fingerprints and keep distinct `NetworkBody` records with correct metadata.

File: packages/playwright-reporter-llm/src/lib/internal/signalFiltering.ts
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/signalFiltering.ts:1]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/1835fdbae9bfb6950671bbc7548772d73d7e83cf. Confirmed via grep that `canImproveConsoleSignal` had no production callers after the short-circuit removal — only the function itself and its tests referenced it. Deleted the function from `signalFiltering.ts` and removed the `describe(canImproveConsoleSignal, ...)` block plus its import from `signalFiltering.test.ts`. No other symbols become unused.

File: packages/playwright-reporter-llm/src/lib/internal/networkBuilder.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/networkBuilder.ts:554]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/1835fdbae9bfb6950671bbc7548772d73d7e83cf. Same root cause as the CodeRabbit finding on this line — swapped `finalize()` to sort before annotating redirects so iteration uses the same offsetMs/instanceSequence key as downstream consumers. Covers the multi-trace interleave scenario you flagged: the new regression test admits the target instance BEFORE the redirecting source in insertion order but gives it a LATER offsetMs, and mutation-verified the test fails under the old ordering.

File: packages/playwright-reporter-llm/src/lib/internal/networkBuilder.test.ts [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+      resourceType: "fetch",
+      ...shape,
+    },
+    instance: { offsetMs: 0, ...instance },
+    ...bodies,
+  };
+}
+
+function makeBody(content: string, truncated = false): NetworkObservationBody {
+  return { content, contentType: "application/json", truncated, fingerprint: hashBody(content) };
```
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/networkBuilder.test.ts]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/5d731643f275df95c478cf755c9b1c2eb1db0f9f. Agreed — updated `makeBody(content, truncated?, contentType?)` to pass the same contentType into both the persisted field and `hashBody(content, contentType)`. Fixtures now match production hashing, so the MIME-collision test (`does not collide when a delimiter-like character is embedded in a field`) and the new `does not collapse bodies with identical bytes but different content types` test both exercise the real dedupe path.

File: packages/playwright-reporter-llm/src/lib/internal/networkBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+    const existingGroupId = this.fingerprintToGroupId.get(fingerprint);
+
+    if (existingGroupId !== undefined) {
+      const metadata = this.shapeMetadata.get(fingerprint);
+      if (metadata) {
+        if (instance.offsetMs !== undefined) {
+          metadata.lastOffsetMs = instance.offsetMs;
+          metadata.firstOffsetMs ??= instance.offsetMs;
+        }
+        metadata.totalPostFilterObservations += 1;
```
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/networkBuilder.ts:160]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/5d731643f275df95c478cf755c9b1c2eb1db0f9f. Agreed — switched the shapeMetadata update to `firstOffsetMs = min(existing, new)` and `lastOffsetMs = max(existing, new)`. That's the correct bookkeeping under multi-trace merge where a later `admit()` can carry an earlier `offsetMs`. The previous assignment logic would (a) freeze `firstOffsetMs` to whatever admit arrived first and (b) overwrite `lastOffsetMs` with the most-recent admit — both wrong when admissions aren't in temporal order. Added a regression test (`takes min(firstOffsetMs)/max(lastOffsetMs) across out-of-order admissions`) and mutation-verified it fails under the old code.

File: packages/playwright-reporter-llm/src/lib/internal/networkBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+    const fingerprint = hashShape(shape);
+    const existingGroupId = this.fingerprintToGroupId.get(fingerprint);
+
+    if (existingGroupId !== undefined) {
+      const metadata = this.shapeMetadata.get(fingerprint);
+      if (metadata) {
+        if (instance.offsetMs !== undefined) {
+          metadata.lastOffsetMs = instance.offsetMs;
+          metadata.firstOffsetMs ??= instance.offsetMs;
+        }
```
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/networkBuilder.ts:159]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/5d731643f275df95c478cf755c9b1c2eb1db0f9f. Same root cause and fix as the CodeRabbit finding on the same block — shapeMetadata now takes min(firstOffsetMs) and max(lastOffsetMs) across admissions, so out-of-order admits from multi-trace merge no longer clobber the true bounds. Added a regression test and mutation-verified it fails under the old assignment logic.

File: packages/playwright-reporter-llm/src/lib/internal/networkProcessing.test.ts
[the remark this answers is not in the document]
Comment (therockstorm) [reply packages/playwright-reporter-llm/src/lib/internal/networkProcessing.test.ts:1]: Addressed in https://github.com/ClipboardHealth/core-utils/commit/54080e58c8b612714f17075a20809f454c7c5b63. Agreed — the integration-level parseTraceparent edge cases were genuinely uncovered after the refactor moved `parseTraceparent` into its own module without moving tests. Added `traceparent.test.ts` with 12 cases covering: well-formed parsing, whitespace/case normalization, undefined/empty input, malformed shape, non-hex characters, all-zero trace id, all-zero span id, `ff` version, truncated input, wrong-length trace id, and wrong-length span id. The `startedDateTime`-fallback and no-timing-source cases remain covered at the helper layer in `traceTiming.test.ts` (lines 58 and 68), so that coverage wasn't actually lost — just relocated to the right unit.
